File: heap/heap.ts
// n 为索引
// 左子节点 2n 右子节点为 2n + 1
// 父节点为 n >> 1

class MaxHeap<T extends {value: number}> {
  private count: number;
  private heap: T[];
  public readonly size?: number;

  constructor(size?: number) {
    this.heap = new Array(size && size + 1);
    this.count = 0;
    this.size = size;
  }

  swap(i: number, j: number) {
    [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]]
  }

  push(item: T) {
    if (this.size !== undefined && this.count === this.size) {
      return false
    }
    this.count++;
    this.heap[this.count] = item;

    let i = this.count;
    while (i > 1) {
      if (this.heap[i].value > this.heap[i >> 1].value) {
        this.swap(i, i >> 1)
        i = i >> 1
      } else {
        break
      }
    }
    return true
  }

  pop() {
    if (this.count === 0) {
      return null
    }

    const item = this.heap[1]
    this.heap[1] = this.heap[this.count];
    this.heap[this.count] = undefined
    let i = 1
    while (this.heap[2 * i]) {
      let maxPos = i;
      if (this.heap[2 * i].value > this.heap[maxPos].value) {
        maxPos = 2 * i
      }
      if (this.heap[2 * i + 1] && this.heap[2 * i + 1].value > this.heap[maxPos].value) {
        maxPos = 2 * i + 1
      }
      if (maxPos === i) {
        break
      } else {
        this.swap(i, maxPos);
        i = maxPos;
      }
    }
    this.count--
    return item
  }
}

class MinHeap<T extends {value: number}> {
  private count: number;
  private heap: T[];
  public readonly size?: number;

  constructor(size?: number) {
    this.heap = new Array(size && size + 1);
    this.count = 0;
    this.size = size;
  }

  swap(i: number, j: number) {
    [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]]
  }

  push(item: T) {
    if (this.size !== undefined && this.count === this.size) {
      return false
    }
    this.count++;
    this.heap[this.count] = item;

    let i = this.count;
    while (i > 1) {
      if (this.heap[i].value < this.heap[i >> 1].value) {
        this.swap(i, i >> 1)
        i = i >> 1
      } else {
        break
      }
    }
    return true
  }

  pop() {
    if (this.count === 0) {
      return null
    }

    const item = this.heap[1]
    this.heap[1] = this.heap[this.count];
    this.heap[this.count] = undefined
    let i = 1
    while (this.heap[2 * i]) {
      let minPos = i;
      if (this.heap[2 * i].value < this.heap[minPos].value) {
        minPos = 2 * i
      }
      if (this.heap[2 * i + 1] && this.heap[2 * i + 1].value < this.heap[minPos].value) {
        minPos = 2 * i + 1
      }
      if (minPos === i) {
        break
      } else {
        this.swap(i, minPos);
        i = minPos;
      }
    }
    this.count--
    return item
  }
}

File: heap/heap.rs
trait HasValue {
    type Value: PartialOrd;
    fn value(&self) -> &Self::Value; 
}

struct MaxHeap<T: HasValue> {
    heap: Vec<T>,
    size: usize,
    count: usize,
}

impl<T: HasValue> MaxHeap<T> {
fn new(size: usize) -> MaxHeap<T> {
    MaxHeap {
        heap: Vec::new(),
        size,
        count: 0,
    }
}

fn swap(&mut self, i: usize, j: usize) {
    self.heap.swap(i, j);
}

fn push(&mut self, item: T) -> bool {
    if self.size <= self.count {
        return false
    }

    self.heap[self.count] = item;
    self.count += 1;

    let mut i = self.count;

    while i > 0 {
        if self.heap[(i - 1) >> 1].value() < self.heap[i].value() {
            self.swap(i, (i - 1) >> 1);
            i = (i - 1) >> 1;
        } else {
            break
        }
    }
    true
}

fn pop(&mut self) -> Option<T> {
    if self.count == 0 {
        return None
    } 

    let item = self.heap.swap_remove(1);
    self.heap[1] = self.heap.swap_remove(self.count);
    self.count -= 1;

    let mut i = 0;
    while i * 2 + 1 <= self.count {
        let mut max_pos = i;
        if self.heap[i].value() < self.heap[i * 2 + 1].value() {
            max_pos = i * 2 + 1;
        }
        if i * 2 + 2 <= self.count && self.heap[i].value() < self.heap[i * 2 + 2].value() {
            max_pos = i * 2 + 2;
        }
        
        if max_pos == i {
            break;
        }
        self.swap(i, max_pos);
        i = max_pos;
    }

    Some(item)
}